doc/tutorials/cpu_sharing.rst

```diff
@@ -0,0 +1,142 @@
+- VM2: 2 vCPUs, pinned to pCPU2 and pCPU3
+.. figure:: images/cpu_sharing_affinity.png
+   :align: center
+
+Thread object state
+*******************
+
+There are three states for thread object, RUNNING, RUNNABLE and BLOCK. States transition happens in certain conditions.
+
+.. figure:: images/cpu_sharing_state.png
+   :align: center
+
+After a new vCPU is created, the corresponding thread object is initiated. The vcpu layer will invoke a wakeup operation. After wakeup, the state for the new thread object is set to RUNNABLE, then follow its algorithm to determine to preempt the current running thread object or not. If yes, it will turn to RUNNING state. In RUNNING state, the thread object may turn back to RUNNABLE state when it runs out of its timeslice, or it yields the pCPU by itself, or be preempted. The thread object under RUNNING state may trigger sleep to transfer to BLOCKED state. 
+
+Scheduler
+*********
+
+Below block diagram shows the basic concept for the scheduler. There are two kinds of scheduler in the diagram, one is NOOP(No-Operation) scheduler and the other is IORR (IO sensitive Round-Robin) scheduler. 
+
+
+- **No-Operation scheduler**:
+
+  The NOOP(No-operation) scheduler has the same policy with original partition mode, every pCPU can run only two thread objects, one is idle thread, and another is the thread of the assigned vCPU. With this scheduler, vCPU work in work-conserving mode, and will run once it’s ready. idle thread could run when the vCPU thread blocked.
+
+- **IO sensitive round-robin scheduler**:
+
+  IORR(IO sensitive round-robin) scheduler is implemented with per-pCPU runqueue and per-pCPU tick timer, it supports more than one vCPU running on a pCPU. The scheduler schedule thread objects in round-robin policy basically, and support preemption by timeslice counting.
+
+  + Every thread object has an initial timeslice (ex: 10ms)
+  + timeslice is consumed with time and be counted in context switch and tick handler
+  + If timeslice is positive or zero, then switch out current thread object and put it to tail of runqueue.Then, pick next runnable one from runqueue to run.
+  + Threads who has IO request will preempt current running thread on the same pCPU.
+
+Scheduler configuration
+***********************
+
+There are two place in the code decide the usage for scheduler.
+
+* The option in Kconfig decides the only scheduler used in runtime.
+  ``hypervisor/arch/x86/Kconfig``
+
+  .. literalinclude:: ../hypervisor/arch/x86/Kconfig
+     :name: Kconfig for Scheduler
+     :caption: Kconfig for Scheduler    
+     :linenos:
+     :lines: 40-58
+     :emphasize-lines: 42
+     :language: c
+
+The default scheduler is **SCHED_NOOP**. To use the IORR, only need to set **SCHED_IORR** in **ACRN Scheduler**. 
+
+* The affinity for VMs are set in  ``hypervisor/scenarios/<scenario_name>/vm_configurations.h``
+
+   .. literalinclude:: ../hypervisor/scenarios/industry/vm_configurations.h
```

Comment:
The literalinclude on line 89 does not properly line up. It should be the same as line 77.

Reply:
Fixed, thanks.